geant4: new version 10.6 plus simplifications

drbenmorgan opened the pull request:
This is a fairly major refactor and simplification of Geant4 and its associated data packages, and *will* changes hashes of previous installs. I think this is worthwhile for long term maintenance of the package, so this PR is "Draft" for now to get early eyes on it and discuss with interested parties.

The primary change is the removal of the `data` variant so the physics model data packages are always installed by the `g4XXX` packages, then used by Geant4 itself through the `geant4-data` package. `geant4-data` is now a `BundlePackage` whose version maps to that of `geant4` and depends on the `g4XXX` packages at the exact versions the `geant4` version supports. It creates a "view" (more on that later) of the data so Geant4's install mechanisms sees a single directory. Each `g4XXX` package also now sets the appropriate runtime environment.

I believe removal of the `data` variant is worthwhile as the external data is platform/arch independent, so really benefits reuse and reducing install sizes.

The second breaking change is to move the versioning to "standard" semantic versioning, so `10.6.0` rather than `10.06.p00`. This is just to make the packaging cleaner and simple, plus the use of the Geant4 GitLab repo.

The remainder of the updates are pure simplification and cleanup. 

One potential issue to look at is the "view" creation in `geant4-data`. All this does is create a directory `<prefix>/share/geant4-data-<version>/` and create symlinks in that to the `g4xxx` data  directories. That works, and Geant4's seems fine with it. However, if a spack environment/view is created and `geant4-data` installed in this, the symlinks under `<viewprefix>/share/geant4-data-<version>/` are created as normal but empty directories.

I'm not sure this is critical as the environment variables needed by Geant4 are set correctly, but feels like I've missed something.

drbenmorgan commented:
It would also be best to wait for @sethrj 's #14520 on vecgeom to be merged so we can take advantage of it for that dependence.

ChristianTackeGSI commented on var/spack/repos/builtin/packages/g4abla/package.py:
```diff
@@ -11,19 +11,24 @@ class G4abla(Package):
         install_tree(self.stage.source_path, install_path)
 
+    def setup_dependent_run_environment(self, env, dependent_spec):
+        install_path = join_path(prefix.share, 'data', 'G4ABLA{0}'
```
Does this actually work?
Where is just `prefix` set?

Could you try this?
```suggestion
    def setup_dependent_run_environment(self, env, dependent_spec):
        install_path = join_path(self.prefix.share.data, 'G4ABLA{0}'
```

drbenmorgan replied:
Fixed in 0d4a05a

ChristianTackeGSI commented on var/spack/repos/builtin/packages/g4incl/package.py:
```diff
@@ -0,0 +1,34 @@
+
+    def url_for_version(self, version):
+        """Handle version string."""
+        return ("http://geant4-data.web.cern.ch/geant4-data/datasets/G4INCL.%s.tar.gz" % version)
```
Is this really needed? Will the normal version interpolation into `url` really fail?

drbenmorgan replied:
Unfortunately yes, due to the use of `.` between the package name and version, combined with the name including a number. Spack interprets the version as `4INCL.X.Y`.

drbenmorgan commented:
Now rebased on `develop` and added an additional `vecgeom` version needed by Geant4 10.5. I just need to test locally and tidy up the options, then this'll be ready for final review/test.

drbenmorgan commented:
O.k., I've done basic install tests and defaults seem to work fine. I think this is ready for final review checks and tests.

ChristianTackeGSI commented on var/spack/repos/builtin/packages/geant4-data/package.py:
```diff
@@ -8,90 +8,89 @@
-        url = url + 'ReleaseNotes4.{0}.{1}.html'.format(version[0], version[1])
-        return url
+            for s in spec.dependencies():
+                for d in glob.glob('{0}/share/data/*'.format(s.prefix)):
```
Maybe use `s.prefix.share` to be more consistent with the rest of the patch?

ChristianTackeGSI commented on var/spack/repos/builtin/packages/geant4/package.py:
```diff
@@ -15,84 +13,110 @@ class Geant4(CMakePackage):
+
+        # install the data with geant4
+        datadir = spec['geant4-data'].prefix.share
+        dataver = 'geant4-data-{0}'.format(spec['geant4-data'].version)
```
Maybe to be consistent with the setup in geant4-data?
```suggestion
        dataver = '{0}-{1}'.format(spec['geant4-data']..name, spec['geant4-data'].version.dotted)
```

Also I wonder, if geant4-data could have a property/value that could be used here?

ChristianTackeGSI commented:
Looking at python's `def command(self)` maybe this can be done on the geant-data class:

```python
    @property
    def geant4_datadir(self):
        verdir = '{0}-{1}'.format(self.name, self.version.dotted)
        return join_path(self.prefix.share, 'data', verdir)
```

UNTESTED!

sethrj commented on var/spack/repos/builtin/packages/geant4/package.py:
```diff
@@ -15,84 +13,110 @@ class Geant4(CMakePackage):
+        ]
 
+        # Multithreading
+        options.append('-DGEANT4_BUILD_MULTITHREADED=' + on_or_off('threads'))
```
With `https://github.com/spack/spack/pull/14376` you can write:
```python
options.append(self.define_from_variant('GEANT4_BUILD_MULTITHREADED', 'threads'))
```

drbenmorgan commented:
Thanks @ChristianTackeGSI and @sethrj for the review! I think I've addressed these in the latest commits, so barring flake8 errors and any further comments, I think this is good for merge.

drbenmorgan commented:
Whilst the coverage checks are failing, the changes in this PR *aren't* (at least obviously) touching Spack's core, which the coverage handles (thanks to @alalazo for the info). I'm going to merge in this case (and I'm maintainer of these packages, so I'll get pinged if there are issues!).
